### PR TITLE
include InitContainers in Kubernetes Service Discovery

### LINKED ISSUE #5563: Kubernetes service discovery for initContainers
## Proposal

The current Kubernetes service discovery implementation totally disregard initContainers.

https://github.com/prometheus/prometheus/blob/master/discovery/kubernetes/pod.go#L216

This would be useful to have them scrape because:
- they may contain troubleshooting information.
- We could detect problems.
- We might want to observe how long initialization took.

Originally the problem appeared from https://github.com/grafana/loki/issues/575  while for metrics it is arguable that you might need this, for logs there is no question.

Alternatives could be to create a custom service discovery. /cc @gouthamve

### PR DESCRIPTION
**Pull request by sh0rez**
@brian-brazil This PR adds support for InitContainers in Kubernetes SD. Closes #5563

Signed-off-by: sh0rez <me@shorez.de>

### CODE REVIEW

**Comment by brian-brazil** [comment]
Is there a metadata label or something to distinguish these?

**Comment by simonpasquier** [comment]
At least we should have a metadata label differentiating regular containers from init containers.

**Comment by sh0rez** [comment]
> @brian-brazil  
> Is there a metadata label or something to distinguish these?


I've added a new metadata label `__meta_kubernetes_pod_container_is_init` to distinguish `InitContainers` from regular ones.  
It takes the values `"true"` if it is an `InitContainer`, `"false"` otherwise.

Tests have been adapted.

**Comment by simonpasquier** [comment]
Can you add a test that where the label is `true`? You'd also need to mention the label in https://github.com/prometheus/prometheus/blob/master/docs/configuration/configuration.md.

**Comment by sh0rez** [comment]
> @simonpasquier 
> Can you add a test that where the label is `true`? You'd also need to mention the label in https://github.com/prometheus/prometheus/blob/master/docs/configuration/configuration.md.

Addressed these

File: discovery/kubernetes/pod.go [hunk 1]
```diff
@@ -213,7 +214,10 @@ func (p *Pod) buildPod(pod *apiv1.Pod) *targetgroup.Group {
 	tg.Labels = podLabels(pod)
 	tg.Labels[namespaceLabel] = lv(pod.Namespace)
 
-	for _, c := range pod.Spec.Containers {
+	containers := append(pod.Spec.Containers, pod.Spec.InitContainers...)
+	for i, c := range containers {
+		is_init := i >= len(pod.Spec.Containers)
```
Comment (simonpasquier) [inline discovery/kubernetes/pod.go]: (nit) s/is_init/isInit/ to follow Go's [conventions](https://golang.org/doc/effective_go.html#mixed-caps).

File: docs/configuration/configuration.md [hunk 2]
```diff
@@ -757,6 +757,7 @@ Available meta labels:
 * `__meta_kubernetes_pod_container_port_name`: Name of the container port.
 * `__meta_kubernetes_pod_container_port_number`: Number of the container port.
 * `__meta_kubernetes_pod_container_port_protocol`: Protocol of the container port.
+* `__meta_kubernetes_pod_container_port_is_init`: Designates whether the container is an [InitContainer](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/)
```
Comment (simonpasquier) [inline docs/configuration/configuration.md]: `__meta_kubernetes_pod_container_is_init` and please keep the list sorted.
I'd rephrase the description to "`true` if the container is an [InitContainer](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/)."

File: docs/configuration/configuration.md [hunk 3]
```diff
@@ -754,6 +754,7 @@ Available meta labels:
 * `__meta_kubernetes_pod_annotation_<annotationname>`: Each annotation from the pod object.
 * `__meta_kubernetes_pod_annotationpresent_<annotationname>`: `true` for each annotation from the pod object.
 * `__meta_kubernetes_pod_container_name`: Name of the container the target address points to.
+* `__meta_kubernetes_pod_container_port_is_init`: `true` if the container is an [InitContainer](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/)
```
Comment (simonpasquier) [inline docs/configuration/configuration.md]: `__meta_kubernetes_pod_container_init`

Comment (sh0rez) [reply docs/configuration/configuration.md]: Adapted

File: docs/configuration/configuration.md [hunk 4]
```diff
@@ -754,6 +754,7 @@ Available meta labels:
 * `__meta_kubernetes_pod_annotation_<annotationname>`: Each annotation from the pod object.
 * `__meta_kubernetes_pod_annotationpresent_<annotationname>`: `true` for each annotation from the pod object.
 * `__meta_kubernetes_pod_container_name`: Name of the container the target address points to.
+* `__meta_kubernetes_pod_container_port_init`: `true` if the container is an [InitContainer](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/)
```
Comment (simonpasquier) [inline docs/configuration/configuration.md]: You still have to remove the `_port` part and move it one line above.

Comment (sh0rez) [reply docs/configuration/configuration.md]: Oh, I'm sorry, I didn't realize this was still there :see_no_evil:

Comment (simonpasquier) [reply docs/configuration/configuration.md]: No problem, this is what code reviews are for :-)

File: discovery/kubernetes/pod.go [hunk 5]
```diff
@@ -141,6 +141,7 @@ const (
 	podContainerPortNameLabel     = metaLabelPrefix + "pod_container_port_name"
 	podContainerPortNumberLabel   = metaLabelPrefix + "pod_container_port_number"
 	podContainerPortProtocolLabel = metaLabelPrefix + "pod_container_port_protocol"
+	podContainerIsInit            = metaLabelPrefix + "pod_container_init"
```
Comment (bwplotka) [inline discovery/kubernetes/pod.go]: Just checked and seems like container names can be not unique which this metadata label must have, nice :+1:

Comment (bwplotka) [inline discovery/kubernetes/pod.go]: Just for the sake of clarity: I thought in this PR the consensus was on `pod_container_is_init` is it a bug or `is` is omitted intentionally? I personally think `pod_container_is_init` makes more sense.
```suggestion
	podContainerIsInit            = metaLabelPrefix + "pod_container_is_init"
```

File: docs/configuration/configuration.md [hunk 6]
```diff
@@ -753,6 +753,7 @@ Available meta labels:
 * `__meta_kubernetes_pod_labelpresent_<labelname>`: `true`for each label from the pod object.
 * `__meta_kubernetes_pod_annotation_<annotationname>`: Each annotation from the pod object.
 * `__meta_kubernetes_pod_annotationpresent_<annotationname>`: `true` for each annotation from the pod object.
+* `__meta_kubernetes_pod_container_init`: `true` if the container is an [InitContainer](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/)
```
Comment (bwplotka) [inline docs/configuration/configuration.md]:
```suggestion
* `__meta_kubernetes_pod_container_is_init`: `true` if the container is an [InitContainer](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/)
```

File: discovery/kubernetes/pod_test.go [hunk 7]
```diff
@@ -128,6 +170,7 @@ func expectedPodTargetGroups(ns string) map[string]*targetgroup.Group {
 					"__meta_kubernetes_pod_container_port_name":     "testport",
 					"__meta_kubernetes_pod_container_port_number":   "9000",
 					"__meta_kubernetes_pod_container_port_protocol": "TCP",
+					"__meta_kubernetes_pod_container_init":          "false",
```
Comment (bwplotka) [inline discovery/kubernetes/pod_test.go]:
```suggestion
					"__meta_kubernetes_pod_container_is_init":          "false",
```
And everywhere else in this file if we agreed on this.

File: discovery/kubernetes/pod.go (hunk 5 above)
Comment (bwplotka) [reply discovery/kubernetes/pod.go]: Ignore me, seen https://github.com/prometheus/prometheus/pull/5598#discussion_r288481795 comment.

File: docs/configuration/configuration.md (hunk 6 above)
Comment (bwplotka) [reply docs/configuration/configuration.md]: ignore this (: Seems like ommiting `is` is what we want. :+1:

File: discovery/kubernetes/pod_test.go (hunk 7 above)
Comment (bwplotka) [reply discovery/kubernetes/pod_test.go]: ignore this (: Seems like ommiting `is` is what we want. :+1:

**Comment by sh0rez** [comment]
@brian-brazil can this one be merged, please? Or are there any blockers I am not aware of? We need this over at [grafana/loki](https://github.com/grafana/loki) as InitContainers are critical for logs.

**Comment by brancz** [comment]
While in this instance it makes sense for prometheus to have this, I'm not sure this will always hold true for things that external projects need. I don't think that's generally a justification.

**Comment by sh0rez** [comment]
> While in this instance it makes sense for prometheus to have this, I'm not sure this will always hold true for things that external projects need. I don't think that's generally a justification.

Of course not! But as this one makes sense for Prometheus and has been approved, I wondered whether it can be merged, to benefit both projects. I just wanted to mention Loki because it was the original motivation for this change :)

**Comment by brancz** [comment]
lgtm :+1:
